### PR TITLE
replace kill() with Signal(syscall.SIGINT) 

### PR DESCRIPTION
**Pull request by matthewgao**
Kill signal will force the process exit, so it won't work in graceful mode

### CODE REVIEW

File: grace/server.go [hunk 1]
```diff
@@ -120,7 +123,7 @@ func (srv *Server) ListenAndServeTLS(certFile, keyFile string) (err error) {
 			log.Println(err)
 			return err
 		}
-		err = process.Kill()
+		err = process.Signal(syscall.SIGINT)
```
Comment (amrfaissal) [inline grace/server.go]: Sending SIGINT is not supported on Windows.

File: grace/server.go [hunk 2]
```diff
@@ -187,11 +191,15 @@ func (srv *Server) handleSignals() {
...
 			log.Printf("Received %v: nothing i care about...\n", sig)
 		}
-		srv.signalHooks(PostSignal, sig)
+		go srv.signalHooks(PostSignal, sig)
+		srv.wg.Done()
 	}
 }
 
 func (srv *Server) signalHooks(ppFlag int, sig os.Signal) {
+	srv.wg.Add(1)
```
Comment (amrfaissal) [inline grace/server.go]: This should be outside of the function

File: grace/server.go [hunk 3]
```diff
@@ -240,6 +248,7 @@ func (srv *Server) serverTimeout(d time.Duration) {
 			break
 		}
 		srv.wg.Done()
```
Comment (amrfaissal) [inline grace/server.go]: Why you're still calling `srv.wg.Done()` when it's already deferred at the beginning.

File: grace/server.go [hunk 4]
```diff
@@ -240,6 +248,7 @@ func (srv *Server) serverTimeout(d time.Duration) {
...
+		runtime.Gosched()
```
Comment (amrfaissal) [inline grace/server.go]: We are already using a `WaitGroup` for that purpose.

File: grace/server.go (hunk 1 above)
Comment (matthewgao) [inline grace/server.go]: But in UNIX like OS Kill will force to kill the process, it won't wait for the parent gracefully exit, if INT is not support on windows then we can change it to another one.

File: grace/server.go (hunk 2 above)
Comment (matthewgao) [inline grace/server.go]: If we don't add the wait group, then we have a change which the process exit before any pre/post handler get called.

File: grace/server.go (hunk 3 above)
Comment (matthewgao) [inline grace/server.go]: I may not catch your point, the defer is used for handling a panic, not decrease a waitGroup

File: grace/server.go (hunk 4 above)
Comment (matthewgao) [inline grace/server.go]: It is a corner case, if we don't goshced here, then we have a change to loop again without srv.state been changed to StateTerminate, this race condition will make the waitGroup step == -1, that's a panic.

Comment (astaxie) [inline grace/server.go]: which race condition? I want to know the root cause here while not use Gosched to ignore the race condition

Comment (matthewgao) [inline grace/server.go]: StateTerminate is not protected by any mutex and not atomic, so the this for loop has a change not jump out, so the Done() will be called additionally, so the 'step' in waitGroup become -1

**Comment by astaxie** [comment]
merged in another PR

**Pull request closed without merging**